elooplib/threadpool.cpp

```diff
@@ -0,0 +1,46 @@
+#include "threadpool.hpp"
+
+ThreadPool::ThreadPool(size_t num_threads)
+{
+  threads.reserve(num_threads);
+  for (size_t i = 0; i < num_threads; i++)
+  {
+    threads.emplace_back(&ThreadPool::run, this);
+  }
+}
+
+void ThreadPool::addTask(std::shared_ptr<ITask> &task)
+{
+  std::lock_guard<std::mutex> q_lock(q_mtx);
+  t_queue.push(task);
+  q_cv.notify_one();
+}
+
+void ThreadPool::run()
```

Comment:
That's a canonical thread pool, very common pattern. Pay attention to understand how it works.